src/libsystemd/sd-event/event-source.h

```diff
@@ -42,6 +43,12 @@ typedef enum WakeupType {
         _WAKEUP_TYPE_INVALID = -1,
 } WakeupType;
 
+typedef struct {
+        RateLimit r;
+        sd_event_source *timeout_source;
+        bool effective;
+} EventSourceRateLimit;
```

Comment:
I think it would be simpler to fold this into the main event source structure. the needed space is 4 64bit values instead 1, it's not too horrible. It feels like premature optimization to me to make this indirect.

We usually turn off ratelimiting of by setting the interval or burst fields to zero. ratelimit_below() recognizes that exlicitly. I'd stick to that same logic here for marking disabled ratelimiting.